File: swift/Sources/allmark/RenderHtml.swift
import Foundation
import Collections

@MainActor
func renderHtml(doc: MarkdownNode, renderers: OrderedDictionary<String, Renderer>) -> String {
	var state = RendererState(
		renderers: renderers,
		output: "",
		footnotes: [],
		depth: 0,
		quoteDepth: 0
	)

	renderChildren(node: doc, state: &state)

	if !state.footnotes.isEmpty {
		renderFootnoteList(state: &state)
	}

	if !state.output.isEmpty && !state.output.hasSuffix("\n") {
		state.output += "\n"
	}

	return state.output
}

File: dotnet/Allmark/RenderHtml.cs
namespace Allmark;

using System.Text;
using Allmark.Render;
using Allmark.Types;

public static class RenderHtml
{
    public static string Execute(MarkdownNode doc, Dictionary<string, Renderer> renderers)
    {
        var state = new RendererState
        {
            Renderers = renderers,
            Output = new StringBuilder(),
            Footnotes = new List<MarkdownNode>(),
            Depth = 0,
            QuoteDepth = 0
        };

        RenderChildren.Execute(doc, state);

        if (state.Footnotes.Count > 0)
        {
            RenderFootnoteList.Execute(state);
        }

        if (state.Output.Length > 0 && state.Output[^1] != '\n')
        {
            state.Output.Append('\n');
        }

        return state.Output.ToString();
    }
}
